### PR TITLE
fix bug: kube-proxy with mode ipvs can not clean up SVC ip rules when SVC is deleted

### LINKED ISSUE #78421: kube-proxy with mode ipvs can not clean up inactive svc ip rules
**What happened**:
kube-proxy with mode ipvs can not clean up inactive svc ip rules, when svc is deleted, the rules about this svc ip is still present.
```
ipvsadm -Ln   # rules still present
```
and the link ```kube-ipvs0``` also have this svc ip
```
ip a s kube-ipvs0   # svc ip still present
``` 
is this a bug, or feature?

**What you expected to happen**:
kube-proxy should keep ip and ipvs rules in sync with 

**Environment**:
- Kubernetes version (use `kubectl version`):
v1.14.2
- OS (e.g: `cat /etc/os-release`):
CentOS Linux release 7.6.1810 (Core)
- Kernel (e.g. `uname -a`):
Linux xx.xx.xx.xx 4.14.71-1.el7.centos.x86_64 #1 SMP Thu Sep 20 18:40:04 CST 2018 x86_64 x86_64 x86_64 GNU/Linux
- Install tools:
binary
- Network plugin and version (if this is a network-related bug):
calico: v3.7.2

### PR DESCRIPTION
**Pull request by wenma**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
kube-proxy with mode ipvs can not clean up inactive SVC ip rules, when SVC is deleted, the rules related to this SVC ip is still present. kube-proxy should keep ip and ipvs rules in sync with SVC.

**Which issue(s) this PR fixes**:
Fixes #78421

**Special notes for your reviewer**:
I find the root cause is the logic error in kube-proxy Proxier.cleanLegacyService func.  the ``` for _, rs := range rsList ``` is unreachable.

see:  [https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go#L1697-L1712](https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go#L1697-L1712)

**Does this PR introduce a user-facing change?**:
```release-note
fix kube-proxy Proxier.cleanLegacyService logic error where deleting the ipvs virtual server
```

### CODE REVIEW

File: pkg/proxy/ipvs/proxier.go [hunk 1]
```diff
@@ -1693,23 +1693,26 @@ func (proxier *Proxier) cleanLegacyService(activeServices map[string]bool, curre
 		if _, ok := activeServices[cs]; !ok {
 			rsList, _ := proxier.ipvs.GetRealServers(svc)
 
-			// If we still have real servers graceful termination is not done
-			if len(rsList) > 0 {
```
Comment (mattjmcnaughton) [inline pkg/proxy/ipvs/proxier.go]: Am I correct that this if statement meant either we never attempted to execute the for loop, or when we did, it was iterating over a list of length 0?

File: pkg/proxy/ipvs/proxier.go [hunk 2]
```diff
@@ -1693,23 +1693,26 @@ func (proxier *Proxier) cleanLegacyService(activeServices map[string]bool, curre
...
-				continue
-			}
-
+			okayToDeleteVS := true
 			// Applying graceful termination to all real servers
 			for _, rs := range rsList {
 				uniqueRS := GetUniqueRSName(svc, rs)
 				// If RS is already in the graceful termination list, no need to add it again
 				if proxier.gracefuldeleteManager.InTerminationList(uniqueRS) {
+					okayToDeleteVS = false
```
Comment (mattjmcnaughton) [inline pkg/proxy/ipvs/proxier.go]: Can you please share more (either as a comment or just responding to this code review) about why its not ok to delete VS if proxier.gracefuldeleteManager.InTerminationList(uniqueRS) is true?

**Review by mattjmcnaughton — COMMENTED** [review]
Thanks for your work here :)

Would it be possible to add a unit test verifying this code is doing what we expect? It seems like in the past, the lack of test coverage allowed us to introduce a bug without catching it during CI.

File: pkg/proxy/ipvs/proxier.go (hunk 2 above)
Comment (wenma) [reply pkg/proxy/ipvs/proxier.go]: if proxier.gracefuldeleteManager.InTerminationList(uniqueRS) is true meant it have some RS exist not deleted,  only when the whole RS of VS is deleted, we should delete the VS done.

File: pkg/proxy/ipvs/proxier.go (hunk 1 above)
Comment (wenma) [reply pkg/proxy/ipvs/proxier.go]: yeah, we never attempted to execute the for loop

**Comment by wenma** [comment]
> Thanks for your work here :)
> 
> Would it be possible to add a unit test verifying this code is doing what we expect? It seems like in the past, the lack of test coverage allowed us to introduce a bug without catching it during CI.

Yeah, I have finished to add a unit test for my PR.

**Review by cmluciano — COMMENTED** [review]
I patched in the test additions only on master and found that they indeed fail and the ipv5 entry still exists. From my standpoint this change looks good but we need an approver like @m1093782566 to give this a once-over.

**Comment by rikatz** [comment]
Hi @wenma 

By the mentioned issue (#78421) this was fixed in #80942

Do you still see the issue, or this PR might be closed?

Even this was solved by another PR, thank you for your time :)

**Comment by andrewsykim** [comment]
@wenma sorry I missed this PR, in fairness it should have been merged over https://github.com/kubernetes/kubernetes/pull/80942 since it was first but I missed it :(. Thank you for reporting the issue.  

/close

**Pull request closed without merging**